### DISCUSSION TITLE
Ruby installation is missing Psych. Ruby yaml extension not compiled

### DISCUSSION THREAD

**Discussion by DirtyF** [rbenv/ruby-build#2012] [Build failures] [answered]
My setup:

- macOS Mojave with ruby 2.3.7 as the default ruby (/usr/bin/ruby)
- `brew install ruby ` install ruby 2.5.3 (/usr/local/opt/ruby/bin/ruby)
- gems are installed in ~/gems/

<details>
<Summary>gem env</Summary>

```
RubyGems Environment:
  - RUBYGEMS VERSION: 2.7.8
  - RUBY VERSION: 2.5.3 (2018-10-18 patchlevel 105) [x86_64-darwin18]
  - INSTALLATION DIRECTORY: /Users/frank/gems
  - USER INSTALLATION DIRECTORY: /Users/frank/.gem/ruby/2.5.0
  - RUBY EXECUTABLE: /usr/local/opt/ruby/bin/ruby
  - EXECUTABLE DIRECTORY: /Users/frank/gems/bin
  - SPEC CACHE DIRECTORY: /Users/frank/.gem/specs
  - SYSTEM CONFIGURATION DIRECTORY: /usr/local/Cellar/ruby/2.5.3_1/etc
  - RUBYGEMS PLATFORMS:
    - ruby
    - x86_64-darwin-18
  - GEM PATHS:
     - /Users/frank/gems
     - /Users/frank/.gem/ruby/2.5.0
     - /usr/local/lib/ruby/gems/2.5.0
     - /usr/local/Cellar/ruby/2.5.3_1/lib/ruby/gems/2.5.0
  - GEM CONFIGURATION:
     - :update_sources => true
     - :verbose => true
     - :backtrace => false
     - :bulk_threshold => 1000
     - "gem" => "--no-document"
  - REMOTE SOURCES:
     - https://rubygems.org/
  - SHELL PATH:
     - /usr/local/opt/llvm@4/bin
     - /Users/frank/.rbenv/shims
     - /Users/frank/.rbenv/bin
     - /Users/frank/gems/bin
     - /usr/local/opt/ruby/bin
     - /usr/local/bin
     - /usr/bin
     - /bin
     - /usr/sbin
     - /sbin
     - ~/bin
     - /Users/frank/.go/bin
     - /usr/local/opt/go/libexec/bin
     - /usr/local/opt/fzf/bin
```

</details>

---

When I run `rbenv install 2.6.0-preview3` the following error occurs on macOS Mojave:

```
/Users/frank/.rbenv/versions/2.6.0-preview3/lib/ruby/2.6.0/yaml.rb:3: warning: It seems your ruby installation is missing psych (for YAML output).
To eliminate this warning, please install libyaml and reinstall your ruby.
The Ruby yaml extension was not compiled.
ERROR: Ruby install aborted due to missing extensions
```
The Psych gem is installed for ruby 2.5.3:

```
otool -L ~/gems/gems/psych-3.0.3/lib/psych.bundle
/Users/frank/gems/gems/psych-3.0.3/lib/psych.bundle:
	/usr/local/opt/ruby/lib/libruby.2.5.dylib (compatibility version 2.5.0, current version 2.5.3)
	/usr/local/opt/libyaml/lib/libyaml-0.2.dylib (compatibility version 3.0.0, current version 3.5.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1252.200.5)
	/usr/lib/libobjc.A.dylib (compatibility version 1.0.0, current version 228.0.0)
```

What is wrong? How do I fix this?

**Comment by hsbt** [comment]
Your pysch extension(~/gems/gems/psych-3.0.3/lib/psych.bundle) was built by 2.5.x.

**Comment by DirtyF** [comment]
@hsbt thanks, by uninstalling psych, rbenv installed ruby 2.6.0-preview3 without errors: 

`Installed ruby-2.6.0-preview3 to /Users/frank/.rbenv/versions/2.6.0-preview3`

**Comment by taliesinpenbardd** [comment]
Same problem here. `gem` command won't respond to anything : 
```/usr/local/Cellar/ruby/2.5.3_1/lib/ruby/2.5.0/yaml.rb:5: warning: It seems your ruby installation is missing psych (for YAML output).
To eliminate this warning, please install libyaml and reinstall your ruby.
Traceback (most recent call last):
	15: from /usr/local/bin/gem:21:in `<main>'
	14: from /usr/local/Cellar/ruby/2.5.3_1/lib/ruby/2.5.0/rubygems/gem_runner.rb:44:in `run'
	13: from /usr/local/Cellar/ruby/2.5.3_1/lib/ruby/2.5.0/rubygems/gem_runner.rb:79:in `do_configuration'
	12: from /usr/local/Cellar/ruby/2.5.3_1/lib/ruby/2.5.0/rubygems/gem_runner.rb:79:in `new'
	11: from /usr/local/Cellar/ruby/2.5.3_1/lib/ruby/2.5.0/rubygems/config_file.rb:202:in `initialize'
	10: from /usr/local/Cellar/ruby/2.5.3_1/lib/ruby/2.5.0/rubygems/config_file.rb:349:in `load_file'
	 9: from /usr/local/Cellar/ruby/2.5.3_1/lib/ruby/2.5.0/rubygems.rb:729:in `load_yaml'
	 8: from /usr/local/Cellar/ruby/2.5.3_1/lib/ruby/2.5.0/rubygems/core_ext/kernel_require.rb:59:in `require'
	 7: from /usr/local/Cellar/ruby/2.5.3_1/lib/ruby/2.5.0/rubygems/core_ext/kernel_require.rb:59:in `require'
	 6: from /usr/local/Cellar/ruby/2.5.3_1/lib/ruby/2.5.0/yaml.rb:6:in `<top (required)>'
	 5: from /usr/local/Cellar/ruby/2.5.3_1/lib/ruby/2.5.0/rubygems/core_ext/kernel_require.rb:59:in `require'
	 4: from /usr/local/Cellar/ruby/2.5.3_1/lib/ruby/2.5.0/rubygems/core_ext/kernel_require.rb:59:in `require'
	 3: from /Users/lucdlb/.rvm/gems/ruby-2.5.1/gems/psych-3.0.3/lib/psych.rb:13:in `<top (required)>'
	 2: from /Users/lucdlb/.rvm/gems/ruby-2.5.1/gems/psych-3.0.3/lib/psych.rb:16:in `rescue in <top (required)>'
	 1: from /usr/local/Cellar/ruby/2.5.3_1/lib/ruby/2.5.0/rubygems/core_ext/kernel_require.rb:59:in `require'
/usr/local/Cellar/ruby/2.5.3_1/lib/ruby/2.5.0/rubygems/core_ext/kernel_require.rb:59:in `require': incompatible library version - /Users/username/.rvm/gems/ruby-2.5.1/gems/psych-3.0.3/lib/psych.bundle (LoadError)
```

`ruby -v`:
`ruby 2.5.3p105 (2018-10-18 revision 65156) [x86_64-darwin18]`

OS : 
Mac OS X Mojave 10.14.1

Any idea how to debug that ?

**Comment by DirtyF** [comment]
@taliesinpenbardd so you can not run `gem unsinstall psych`? Maybe try to remove `/Users/username/.rvm/gems/ruby-2.5.1/gems/psych-3.0.3/` manually?

**Comment by taliesinpenbardd** [comment]
@DirtyF I've finally uninstalled RVM, and tried my luck with RBENV. No luck so far, I still can't use the gems I've installed (and reinstalled). I suppose it's a wrong installation location on the system, or perhaps a permissions problem. Still looking, but I guess it's out of the focus of this ticket.
Thanks anyway.

**Comment by DirtyF** [comment]
@taliesinpenbardd FWIW you can choose to store your gems in your home directory, by adding the following to your shell config file:

```
export GEM_HOME=$HOME/gems
export PATH=$HOME/gems/bin:$PATH
```

**Comment by taliesinpenbardd** [comment]
@DirtyF Thank you very much, it seems to work. (and thank you also, I did discover Jekyll and am gonna use it sometime in the future. 😄 )

**Comment by farisarie** [comment]
> 

i have a problem when i trying to installing feedkit library and this warning showup I didn't find any solution, do u know what I should do?

/System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/yaml.rb:3: warning: It seems your ruby installation is missing psych (for YAML output).
To eliminate this warning, please install libyaml and reinstall your ruby.
/Library/Ruby/Site/2.6.0/rubygems/core_ext/kernel_require.rb:85:in `require': dlopen(/Library/Ruby/Gems/2.6.0/gems/psych-4.0.2/lib/psych.bundle, 0x0009): tried: '/Library/Ruby/Gems/2.6.0/gems/psych-4.0.2/lib/psych.bundle' (mach-o file, but is an incompatible architecture (have 'arm64', need 'x86_64')), '/usr/lib/psych.bundle' (no such file) - /Library/Ruby/Gems/2.6.0/gems/psych-4.0.2/lib/psych.bundle (LoadError)

**Comment by eertmanhidde** [comment]
Running `gem uninstall psych` from my older ruby version(2.7.2) fixed the issue for me

**Comment by askdesigners** [comment]
I get this when trying to uninstall 
`Gem psych-4.0.3 cannot be uninstalled because it is a default gem`

**Reply by jasonperrone** [reply]
Have you been stuck on this since July 2022, or did you get past it?

**Reply by theDRB123** [reply]
How did you get past this ? I am stuck with the same thing

**Comment by DavidRawk-Blake** [comment]
I'm trapped by this as well. ASDF seems to have the same problem

**Comment by caged** [comment]
I was setting up a clean machine on MacOS Ventura 13.0 and ran into this issue.  
Running `brew install libyaml` allowed the `rbenv install` to complete without error.

**Reply by kurko** [reply]
This didn't fix it for me, despite uninstalling it, unlinking and relinking and all the incantations possible.

**Reply by sekitaka** [reply]
It works for me!

**Comment by arg** [comment]
Got this issue on Ubuntu 22.04 while installing Ruby 3.2.0 as well. Fixed by installing `libyaml`:
```sh
sudo apt install libyaml-dev
```

**Reply by HLFH** [reply]
```
sudo apt install libyaml-dev
```

This did the trick. Thanks! Was not sure which libyaml* package to install.

**Reply by arknamal** [reply]
ty. worked for me on same ubuntu. I m bcz we are ;)

**Reply by nikolaz90** [reply]
thanks you :)

**Reply by mralaminahamed** [reply]
thank you. this is work for me.

**Reply by aecordova** [reply]
this worked for me!!! :)

**Reply by stratigos** [reply]
Also found this on SO: https://stackoverflow.com/questions/77082389/ubuntu-unable-to-locate-package-libyaml - to confirm for us Ubuntu folks.

**Reply by diegobentes** [reply]
works for me! ty

**Comment by rafaelfranca** [comment]
I get this no matter which version of Ruby I try to install using debian. `libyaml-dev` is already installed, but I can't install any Ruby version because I get the "It seems your ruby installation is missing psych (for YAML output)" error.

Here is the log

https://gist.github.com/rafaelfranca/91250c4079cb483c9cc6796c6356cd82

**Reply by HLFH** [reply]
@rafaelfranca Try this script to clean gems? https://gist.github.com/IanVaughan/2902499

**Reply by kurko** [reply]
Getting this suddenly for all versions as well, but on Mac M2 13.1. Currently installed versions show this,

```
incompatible library version - /Users/alex/.gem/ruby/2.7.0/gems/psych-5.0.2/lib/psych.bundle (LoadError)
/Users/alex/.rbenv/versions/2.7.1/lib/ruby/2.7.0/yaml.rb:3: warning: It seems your ruby installation is missing psych (for YAML output).
To eliminate this warning, please install libyaml and reinstall your ruby.
```

and I can't `rbenv install` any version

**Comment by lunohodov** [comment, marked as answer]
Make sure your environment meets the requirements as [suggested](https://github.com/rbenv/ruby-build/wiki#suggested-build-environment) in the wiki. Then, depending on your system, pass the libyaml path to the installation process.

For me, on an M2 Ventura 13.2.1, with a Homebrew-installed `asdf` (ruby-build under the hood):

```shell
~ % brew reinstall libyaml
~ % gem install psych -- --with-libyaml-dir=$(brew --prefix libyaml)
~ % export RUBY_CONFIGURE_OPTS=--with-libyaml-dir=$(brew --prefix libyaml)
~ % asdf install ruby latest
```

**Reply by mislav** [reply]
@lunohodov Thanks! With the [latest version of ruby-build](https://github.com/rbenv/ruby-build/releases/tag/v20230306), manually linking to Homebrew's libyaml isn't necessary anymore, since ruby-build will automatically pass `--with-libyaml-dir=$(brew --prefix libyaml)` if libyaml was found. 🚀

**Comment by QMiqTx6DHn1bA9yaNaAbsD3CLG8gTmd4** [comment]
@mislav Any ideas why/what could cause the error? 

```
/Users/username/.rubies/2.7.7/lib/ruby/2.7.0/yaml.rb:3: warning: It seems your ruby installation is missing psych (for YAML output).
To eliminate this warning, please install libyaml and reinstall your ruby.
The Ruby yaml extension was not compiled.
ERROR: Ruby install aborted due to missing extensions
```

Full log attached - [1.log](https://github.com/rbenv/ruby-build/files/11081676/1.log)

---

Seemingly it happened already after the Ruby 2.7.7 was successfully installed. Having checked the installation I don't see any issues with YAML/Psych & smoke test code below runs perfectly. 🤷‍♂️ 

```
$ ruby -e "require 'psych'; require 'yaml'; puts YAML.dump('abc')"
--- abc
```

**Reply by mislav** [reply]
I can't spot anything that went wrong in your installation, and I'm puzzled by the error since it looks like Ruby had linked to Homebrew libyaml successfully and that psych worked. Note, however, that Ruby 2.7 is becoming EOL end of this month.

@hsbt: any ideas?

**Reply by hsbt** [reply]
```
installing default gems from ext:   /Users/username/.rubies/2.7.7/lib/ruby/gems/2.7.0 (build_info, cache, doc, extensions, gems, specifications)
6048   │                                     bigdecimal 2.0.0
(snip)
6058   │                                     psych 3.1.0
```

Installation is successful. This user may use `GEM_HOME` and mixed psych binary of several versions.

**Reply by cwatson-im** [reply]
Thanks! I had the same issue in this subthread, and `unset GEM_HOME` fixed it for me.

**Comment by nkokkos** [comment]
My solution:

I use **Rocky Linux 8 ( forked from centos)**  and I tried to install ruby 3.2.2 using:

`rbenv install 3.2.2`

So, I got this message: 

```
Leaving directory '/tmp/ruby-build.20230427112055.48112.QYerKI/ruby-3.2.2'
Generating RDoc documentation
/tmp/ruby-build.20230427112055.48112.QYerKI/ruby-3.2.2/lib/yaml.rb:3: warning: It seems your ruby installation is missing psych (for YAML output). To eliminate this warning, please install libyaml and reinstall your ruby.
```

How I resolved it:

as root, I installed libyaml + libyaml-devel:

`yum install libyaml`

But **_you have to enable powertools repo_** to install source version libyaml-devel:

`yum --enablerepo=powertools install libyaml-devel`

After installing both, at my root prompt I typed;

`ldconfig -vvv`

Finally, I logged in as normal user and installed ruby version 3.2.2:

`rbenv install 3.2.2`

It worked for me.

**Reply by geanttechnology** [reply]
it worked for me as well. Rocky Linux 8 needs libyaml-devel to compile ruby 3.2.2. Thanks

**Reply by amruali** [reply]
Worked for me, Thanks Nick.

**Comment by initseis** [comment]
For Mac M1 with Monterey I could install Ruby 2.3.1 and 3.2.1 with Rosetta following these tutorials:
- https://soffes.com/blog/homebrew-on-apple-silicon
- https://soffes.com/blog/ruby-on-apple-silicon

**Comment by npickens** [comment]
Oof, this was frustrating. A simple 3.1.2 to 3.2.1 Ruby upgrade on CentOS Stream 9 turned into an hours-long event because of this issue. Key thing ended up being enabling `crb` rather than `powertools` on CentOS:

```zsh
dnf config-manager --set-enabled crb
yum install -y libyaml-devel
```

I think the above should do it, but if not, here's everything that I did:

```
# These might not be necessary
yum install -y dnf-plugins-core
yum install -y epel-release

dnf config-manager --set-enabled crb
yum install -y libtool libyaml-devel

# Get location of psych gem and manually delete it (cannot gem uninstall because it's a default gem)
gem info psych
rm -rf <directory where psych is installed>
gem install psych

rbenv install <desired version>
```
